### ISSUE TITLE
Conjugate Gradient (and perhaps eventually Hessian Free) Optimization

### ISSUE THREAD

**Issue by neosatrapahereje** [Lasagne/Lasagne#466]
Hi!

I've implemented Conjugate Gradient for Lasagne, and @f0k asked me to put it online so that we can discuss how to integrate it to Lasagne:

https://gist.github.com/neosatrapahereje/b7427ab63dfc0510cd7b

**Comment by f0k** [comment]
Great, thanks! To get others up to speed: The main goal would be to have a working Hessian-free implementation done completely in Theano. This requires a Conjugate Gradient implementation with backtracking. What Carlos implemented in that gist is something following the interface of `lasagne.updates.*`, with a little twist: The update dictionary doesn't just do a single step, but uses `theano.scan` to run Conjugate Gradient until convergence and return the end result.

While discussing the implementation with Carlos, I wondered if this should split up a bit further. I was thinking of something like:
* `lasagne.updates.cg_step`: Return an update dictionary for a single CG step
* `lasagne.updates.until_convergence`: Use `theano.scan` to apply some update repeatedly until either a given stopping criterion or a maximum number of iterations is reached, optionally with back-tracking with a second loss
* `lasagne.updates.cg`: Convenience function to apply `cg_step` until convergence with a suitable stopping criterion, optionally with back-tracking
* `lasagne.updates.hf`: Implementation of Hessian-free that uses `cg_step` `until_convergence` by default, but could use something else

Separating out `cg_step` and `until_convergence` would allow to track progress of CG while it's running, and allow `until_convergence` to be used with other optimizers and stopping criteria. In theory, you could implement mini-batched SGD in a single scan (but probably that won't buy you anything).

What we need to think through and decide:

1. Do we want this, and is this a good separation? (I think yes and yes.)
2. What exactly would the interface of the different functions look like? (`until_convergence` would probably be the most complex, as it may need some callback for defining the stopping criterion expression. `hf` would also be complex, as it usually uses separate streams of mini-batches for the outer loop and the inner (cg) loop.)

**Comment by benanne** [comment]
This is cool, but is it a common enough use case for this to live in the main library? How popular is CG right now for training neural nets? This is an honest question, I really have no idea. HF seems to have fallen out of favour lately, at least ...

**Comment by craffel** [comment]
> While discussing the implementation with Carlos, I wondered if this should split up a bit further. I was thinking of something like:

That seems reasonable to me.

> This is cool, but is it a common enough use case for this to live in the main library? How popular is CG right now for training neural nets? This is an honest question, I really have no idea. HF seems to have fallen out of favour lately, at least ...

I think as long as people are willing to work on it, it would fit in.  I think there was enough work using HF to train networks that it's worth including.  As far as I know, the main motivation for including a CG implementation would be to implement HF.

**Comment by benanne** [comment]
Sounds fair :) Note that there is already a fairly generic theano HF implementation floating around: https://github.com/boulanni/theano-hf
might be useful as a source of inspiration!

**Comment by f0k** [comment]
> How popular is CG right now for training neural nets? This is an honest question, I really have no idea. HF seems to have fallen out of favour lately, at least ...

CG isn't useful for nonlinear problems, so it would probably only be used within HF, or other to-be-invented optimizers. For HF, I guess one of the main problems is that it's terribly complicated to get right (there are some nifty details), and most things seem to work fine with SGD+Momentum anyway. If it was easy to try HF, more people might do it (even if it was just to discover that it doesn't give an edge over SGD with carefully-chosen initializations).

> Note that there is already a fairly generic theano HF implementation floating around:

Yes, but it's using Theano only for computing the gradient and Rop, everything else happens on CPU in numpy, so it's limited in performance.

**Comment by Botev** [comment]
Also if you need other than HF methods (such as LBFGS for instance) and you want the implemented in Theano you still would need a CG method.

**Comment by nouiz** [comment]
About the generic HF in Theano, yes, it don't do everything in Theano. But
at the time it was done, (I didn't do it, but I talked about this with the
author of this code), the part that Theano wasn't good at was done in
Python.

Using scan in while loop isn't perfect. We don't support is as much as a
number of iteration fixed when the loop start. Also, there was some ifelse
in the graph that caused problems. Theano solution to ifelse isn't perfect
too. So before moving more stuff in Theano and expecting that it would just
be better, take the time to assert that. Theano and in particular scan got
big upgrade since this code, but it isn't certain we would fit the case
needed here.

**Comment by Botev** [comment]
I'm not too sure, but I've read somewhere (or maybe a discussion on reading group) that for HF implementation you can in fact fix the number of CG steps (e.g. you might be ok even if it does not converge on the current iteration). Another option if theano can not make too better, is just to write the Method purely in CUDA and provide an Op which gives you the next step accordingly. And well in the end we don't know if it is going to be slower unless we try.

**Comment by f0k** [comment]
> Using scan in while loop isn't perfect. We don't support is as much as a number of iteration fixed when the loop start. Also, there was some ifelse in the graph that caused problems. Theano solution to ifelse isn't perfect too. So before moving more stuff in Theano and expecting that it would just be better, take the time to assert that.

Thanks for these cautious notes! So before moving on, we should check whether the pure-Theano CG is competitive with doing the loop in Python.

Regarding the main issue, we found that splitting `cg()` into `cg_step()` and `until_convergence()` won't work. There are some variables that need to be initialized before running CG and are then updated at each step. In the current implementation, these variables are symbolic expressions that are updated as part of the scan. If we move out the update into `cg_step()`, they'd need to become shared variables similar to how `momentum()` etc. initialize some shared variables to be used as accumulators of some sorts. The initial value for these variables depends on the gradient of the loss function, though, they're not simply set to zero. So we'd either need to call `eval()` to get the value to initialize the shared variable with (which would cause a compilation, that's something we should avoid in `lasagne.updates.*`), or initialize it to zero and special-case the very first update (which might slow things down). And even if we got this right, it would make it more difficult to use CG as part of another optimization routine such as HF, which would need to reset things correctly before each CG run.

Long story short: We'll probably be better off not splitting up `cg()`.
